kafkakewl-metrics/src/main/scala/com/mwam/kafkakewl/metrics/services/ConsumerOffsetsSource.scala

```diff
@@ -250,20 +267,34 @@ object ConsumerOffsetsSource {
         ZIO
           .scoped {
             for {
-              consumer <- KafkaConsumerUtils.kafkaConsumerBytesBytesZIO(kafkaClientConfig)
-              compactedConsumeResult <- ZIO.attempt(KafkaConsumerUtils.consumeCompactUntilEnd(consumer, topicPartitions))
+              consumer <- Consumer.make(
+                ConsumerSettings(kafkaClientConfig.brokersList, kafkaClientConfig.additionalConfig).withOffsetRetrieval(
+                  Consumer.OffsetRetrieval.Auto(AutoOffsetStrategy.Earliest)
+                )
+              )
+              _ <- ZIO.log(s"Topic partitions asigned to index {index}: " + topicPartitions.mkString(","))
+              compactedConsumeResult <- KafkaConsumerUtils
+                .consumeCompactUntilEnd(
+                  consumer,
+                  topicPartitions,
+                  Serde.bytes,
+                  Serde.bytes
+                )
             } yield compactedConsumeResult
           }
           .timed
           .tap { case (duration, ccr) =>
             ZIO.logInfo(
-              s"#$index consumed ${ccr.noOfConsumedMessages} messages, ${ccr.lastValues.size} unique keys in ${duration.toMillis / 1000.0} seconds (${ccr.noOfConsumedMessages / (duration.toSeconds.toDouble)} messages/sec)"
+              s"#$index consumed ${ccr.noOfConsumedMessages} messages, ${ccr.lastValues.size} unique keys in ${duration.toMillis / 1000.0} seconds (${ccr.noOfConsumedMessages / duration.toSeconds.toDouble} messages/sec)"
             )
           }
-          .map { case (_, ccr) =>
+          .flatMap { case (_, ccr) =>
             ccr.deserialize(
               consumerOffsetsSourceConfig.consumerOffsetsTopicName,
-              ConsumerOffsetsDeserializers.keyDeserializer,
+              ConsumerOffsetsDeserializers.keyDeserializer.mapM {
+                case Some(key) => ZIO.succeed(key)
+                case None      => ZIO.fail(new Exception("Key should not be null"))
```

Comment:
Wouldn't it be clearer if we left the key as an option and call the `CompactedConsumeResult.deserialize` with that?

I think it works with the current implementation but switching from the option to the error type to indicate the that the key should be ignored is a bit misleading and inconsistent with how we handle it at the live consumption above.

This `ZIO.fail(new Exception("Key should not be null"))` above is definitely not ideal because the key can be none and that's perfectly normal.

Reply:
This has been done, good suggestion!